### PR TITLE
PaddleScience 领域经典论文复现 Deep Spatio-Temporal 

### PR DESCRIPTION
**Pull request by co63oc**
### PR types

Others
### PR changes

Others
### Describe

https://github.com/PaddlePaddle/Paddle/issues/55663#paddlepaddle14
图片

Test Loss 信息
| Test Loss | torch | paddle |
| - | - | - |
| wind_power | 1.5525 | 1.5549 |
| wind_speed | 0.2646 | 0.2522 |

wind_power 运行
torch
图片

paddle
图片

wind_speed 运行
torch
图片

paddle
图片

### CODE REVIEW

**Comment by luotao1** [comment]
@njjtan @172w 两位选手请注意下，赛题14的论文二已经提交PR了

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
PR描述里提供的图片看起来好像没有训练完？麻烦提供一下复现的精度和原论文的精度指标

File: jointContribution/Deep-Spatio-Temporal/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+原仓库 https://github.com/jiangyuan-li/Deep-Spatio-Temporal.git
+修改 torch 库为 paddle
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/README.md]: 1

File: jointContribution/Deep-Spatio-Temporal/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+运行 getNRELdata.py 需要在 https://developer.nrel.gov/signup/ 注册，获取API Key，
+然后配置 h5pyd，使用 hsconfigure 命令配置（https://github.com/HDFGroup/hsds_examples）
+endpoint: https://developer.nrel.gov/api/hsds
+------------------------------
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/README.md]: 1

File: jointContribution/Deep-Spatio-Temporal/README.md
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/README.md:1]: 1

File: jointContribution/Deep-Spatio-Temporal/getNRELdata.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/getNRELdata.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 4]
```diff
@@ -0,0 +1,320 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+import numpy as np
+import paddle
+import pandas as pd
+from paddle.io import Dataset
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        file_path="../../data/Wind Spatio-Temporal Dataset2.csv",
+        train_len=60 * 24,
+        val_len=30 * 24,
+        time_len=365 * 24,
+        ENC_LEN=48,
+        DEC_LEN=12,
+        K=5,
+        similarity="spatio",
+    ):
+
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
+#!/usr/bin/env python
+
+import numpy as np
+import paddle
+import pandas as pd
+from paddle.io import Dataset
+from sklearn import preprocessing
+from sklearn.neighbors import NearestNeighbors
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+        time_len=365 * 24,
+        ENC_LEN=48,
+        DEC_LEN=12,
+        K=5,
+        similarity="spatio",
+    ):
+
+        super().__init__(file_path=file_path, K=K)
+
+        self.ids = paddle.to_tensor(paddle.arange(200), dtype=paddle.float32).reshape(
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+            (200, 1)
+        )
+        if K <= 1:
+            self.K = 1
+            self.neighbors = paddle.to_tensor(paddle.arange(200)).reshape((200, 1))
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+
+        self.df_wind_speed = pd.read_csv(folder_path + file_path, index_col=0)
+        self.df_meta = pd.read_csv(folder_path + meta_path, index_col=0)
+        self.wind_speed_tensor = paddle.to_tensor(
+            self.df_wind_speed.to_numpy(), dtype=paddle.float32
+        )
+        self.time_series = pd.to_datetime(
+            [x.split("'")[1] for x in self.df_wind_speed.columns]
+        )
+        #         self._reformat_1d_to_2d()
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+
+class NRELwpDataset(Dataset):
+    def __init__(self, data, ENC_LEN=48, DEC_LEN=12, K=9):
+        self.data = data
+        self.enc_len = ENC_LEN
+        self.total_len = ENC_LEN + DEC_LEN
+        self.K = K
+
+    def __getitem__(self, index):
+
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/datamgr.py [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+        tim = index // 100
+        turbine = index % 100
+        one_point = self.data[turbine, tim : (tim + self.total_len), :]
+        x = one_point[: self.enc_len, :]
+        y = one_point[(self.enc_len - 1) : self.total_len, :]
+
+        # y = y[:, [0, self.K, self.K + 1, self.K + 2]]
```
[the remark this answers is not in the document]
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/datamgr.py]: 1

**Comment by co63oc** [comment]
> PR描述里提供的图片看起来好像没有训练完？麻烦提供一下复现的精度和原论文的精度指标

已增加 Test Loss

**Comment by co63oc** [comment]
> > > PR描述里提供的图片看起来好像没有训练完？麻烦提供一下复现的精度和原论文的精度指标
> > 
> > 
> > 已增加 Test Loss
> 
> 这个好像不是完整训练完的loss？模型复现需要提供完整训练后的指标

图片已更新为训练完成图片
图片

File: jointContribution/Deep-Spatio-Temporal/src/model.py [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+                enc_dim,
+                dec_dim,
+                weight_attr=utils.get_weight_attr(),
+                bias_attr=utils.get_bias_attr(),
+            )
+
+    def forward(self, one_batch):
+        """
+        input_batch: enc_len * size * input_dim
+        """
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/src/model.py:92]: docstring不正确，修改一下，风格与ppsci内代码保持一致

File: jointContribution/Deep-Spatio-Temporal/src/model.py [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+                    self.dec_input_dim,
+                    weight_attr=utils.get_weight_attr(),
+                    bias_attr=utils.get_bias_attr(),
+                )
+
+    def forward(self, one_batch, encoder_outputs, hidden):
+        """
+        input: size * input_dim
+        batch_ids: size
+        """
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/src/model.py:196]: 同上

File: jointContribution/Deep-Spatio-Temporal/src/model.py [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+                )
+                rnn_input = rnn_input.unsqueeze(0)
+
+            hidden = hidden.unsqueeze(0)
+            if self.GRU_LSTM == "GRU":
+                output, hidden = self.rnn(rnn_input, hidden)
+            elif self.GRU_LSTM == "LSTM":
+                output, (hidden, cell_state) = self.rnn(rnn_input, (hidden, cell_state))
+
+            assert (output == hidden).all()
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/src/model.py:233]: 这里的assert是否有必要？一般很少对densetensor进行逐元素检查，否则会占用一定的计算时间吧？

File: jointContribution/Deep-Spatio-Temporal/src/utils.py [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+def get_weight_attr():
+    return paddle.framework.ParamAttr(
+        initializer=paddle.nn.initializer.Normal(mean=0, std=0.1)
+    )
+
+
+def get_bias_attr():
+    return paddle.framework.ParamAttr(
+        initializer=paddle.nn.initializer.Constant(value=0)
+    )
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/src/utils.py:33]: 使用`paddle.ParamAttr`即可

File: jointContribution/Deep-Spatio-Temporal/train.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    parser.add_argument("--batch_size", default=20000, type=int, help="batch size")
+    parser.add_argument("--lr", default=0.001, type=float, help="learning rate")
+    parser.add_argument("--k", default=5, type=int, help="number of spatio neighbors")
+    parser.add_argument(
+        "--n_turbines", default=200, type=int, help="number of turbines"
+    )
+
+    args = parser.parse_args()
+
+    print("Running with following command line arguments: {}".format(args))
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/train.py:29]: 改为f-string

File: jointContribution/Deep-Spatio-Temporal/train.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    K = args.k
+    SAVE_FILE = args.name
+
+    if args.name == "wind_power":
+        if not os.path.exists("./data/wind_power.csv"):
+            sys.exit(
+                "No data found!!!\n"
+                "Download wind power data first (follow the instructions in readme)."
+            )
+        data_mgr = datamgr.DataMgr(file_path="./data/" + args.name + ".csv", K=K)
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/train.py:43]: 路径拼接使用os.path.join

File: jointContribution/Deep-Spatio-Temporal/src/trainer.py [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+            print(f"Val MAE: {np.array(val_mae).reshape(2,6)}")
+            print(f"Val RMSE: {np.array(val_rmse).reshape(2,6)}")
+
+            early_stopping(val_loss)
+            print(f"Best Val Loss: {early_stopping.best_loss:.4f}")
+
+            if early_stopping.early_stop:
+                paddle.save(
+                    self.model.state_dict(), "outputs/" + self.SAVE_FILE + ".pdparams"
+                )
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/src/trainer.py:104]: 路径拼接用os.path.join

File: jointContribution/Deep-Spatio-Temporal/src/trainer.py [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+                break
+        else:
+            paddle.save(
+                self.model.state_dict(), "outputs/" + self.SAVE_FILE + ".pdparams"
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/src/trainer.py:108]: 路径拼接用os.path.join

File: jointContribution/Deep-Spatio-Temporal/train.py [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+        optimizer=optimizer,
+        criterion=criterion,
+        SAVE_FILE=SAVE_FILE,
+        BATCH_SIZE=BATCH_SIZE,
+        name=args.name,
+    )
+    trainer_obj.train(epochs=EPOCHS)
+    loss, mae, rmse = trainer_obj.report_test_error()
+    outdf = pd.DataFrame({"MAE": mae, "RMSE": rmse}, index=[*range(1, 13)])
+    outdf.to_csv("outputs/" + SAVE_FILE + "_metrics.csv")
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/train.py:74]: 路径拼接用os.path.join

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

**Comment by HydrogenSulfate** [comment]
> > > > PR描述里提供的图片看起来好像没有训练完？麻烦提供一下复现的精度和原论文的精度指标
> > > 
> > > 
> > > 已增加 Test Loss
> > 
> > 
> > 这个好像不是完整训练完的loss？模型复现需要提供完整训练后的指标
> 
> 图片已更新为训练完成图片 图片

哦哦好吧，看起来是这个案例收敛速度太快，总的300轮在10轮就提前结束了

File: jointContribution/Deep-Spatio-Temporal/src/model.py (hunk 14 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/model.py]: 已修改增加Args

File: jointContribution/Deep-Spatio-Temporal/src/model.py (hunk 15 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/model.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/trainer.py (hunk 21 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/trainer.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/model.py (hunk 16 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/model.py:243]: 已有数据占用时间不大，如果数据有错误，assert可以验证

File: jointContribution/Deep-Spatio-Temporal/src/utils.py (hunk 17 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/utils.py]: 1

File: jointContribution/Deep-Spatio-Temporal/train.py (hunk 18 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/train.py]: 1

File: jointContribution/Deep-Spatio-Temporal/train.py (hunk 19 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/train.py]: 1

File: jointContribution/Deep-Spatio-Temporal/src/trainer.py (hunk 20 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/trainer.py]: 1

File: jointContribution/Deep-Spatio-Temporal/train.py (hunk 22 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/train.py]: 1

**Comment by co63oc** [comment]
> 哦哦好吧，看起来是这个案例收敛速度太快，总的300轮在10轮就提前结束了

轮次是Epoch，count 这是计算best_loss如果连续10轮小于val_loss，就结束计算

File: jointContribution/Deep-Spatio-Temporal/src/model.py [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+            )
+
+    def forward(self, one_batch):
+        """The forward function of the encoder.
+
+        Args:
+            one_batch (Tensor): The input batch.
+        Returns:
+            Tensor: The encoder outputs.
+            Tensor: The hidden variable returned by the decoder.
```
Comment (HydrogenSulfate) [inline jointContribution/Deep-Spatio-Temporal/src/model.py:96]: 返回值用`Tuple[Tensor, Tensor]`类型

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
返回值类型稍微修改下，其他应该没啥问题了

File: jointContribution/Deep-Spatio-Temporal/src/model.py (hunk 23 above)
Comment (co63oc) [reply jointContribution/Deep-Spatio-Temporal/src/model.py]: 1

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
